cachito/workers/scm.py

```diff
@@ -91,36 +99,74 @@ def repo_name(self):
     """The git implementation of interacting with source control."""
 
+    def _reset_git_head(self, repo):
+        """Reset HEAD to a specific ref
+
+        :param repo: the repository object.
+        :type repo: :class:`git.Repo`
+        :raises CachitoError: if fail to change the HEAD.
+        """
+        try:
+            repo.head.reference = repo.commit(self.ref)
+            repo.head.reset(index=True, working_tree=True)
+        except:  # noqa E722
+            log.exception('Checking out the Git ref "%s" failed', self.ref)
+            raise CachitoError(
+                'Checking out the Git repository failed. Please verify the supplied reference '
+                f'of "{self.ref}" is valid.'
+            )
+
+    def _create_archive(self, from_dir):
+        """Create an archive from a specified directory
+
+        :param str from_dir: path to a directory from where to create the archive.
+        """
+        log.debug('Creating the archive at %s', self.archive_path)
+        with tarfile.open(self.archive_path, mode='w:gz') as bundle_archive:
+            bundle_archive.add(from_dir, 'app')
+
     def clone_and_archive(self):
         """
         Clone the git repository and create the compressed source archive.
 
         :raises CachitoError: if cloning the repository fails or if the archive can't be created
         """
         with tempfile.TemporaryDirectory(prefix='cachito-') as temp_dir:
-            clone_path = os.path.join(temp_dir, 'repo')
             log.debug('Cloning the Git repository from %s', self.url)
             # Don't allow git to prompt for a username if we don't have access
             os.environ['GIT_TERMINAL_PROMPT'] = '0'
             try:
+                clone_path = os.path.join(temp_dir, 'repo')
                 repo = git.repo.Repo.clone_from(self.url, clone_path, no_checkout=True)
             except:  # noqa E722
                 log.exception('Cloning the Git repository from %s failed', self.url)
                 raise CachitoError('Cloning the Git repository failed')
 
+            self._reset_git_head(repo)
+            self._create_archive(repo.working_dir)
+
+    def pull_and_archive(self, previous_archive):
+        """
+        Pull latest git history inside an existing archive and create the
+        compressed source archive
+
+        :param str previous_archive: path to an archive file created before.
+        :raises CachitoError: if fail to pull Git history from remote repo or
+            if fail to checkout to a specific ref.
+        """
+        with tempfile.TemporaryDirectory(prefix='cachito-') as temp_dir:
+            with tarfile.open(previous_archive, mode='r:gz') as tar:
+                tar.extractall(temp_dir)
+
+            repo = git.Repo(os.path.join(temp_dir, 'app'))
             try:
-                repo.head.reference = repo.commit(self.ref)
-                repo.head.reset(index=True, working_tree=True)
+                repo.remote().pull()
```

Comment:
I think the proper approach is to run `git fetch` rather than pull.

One approach is the `update` method:
https://gitpython.readthedocs.io/en/stable/reference.html#git.remote.Remote.update

Reply:
`git-fetch` just fetches the changes but not merge them into local repository. Creating an archive from local repository should require the merge.

Reply:
@tkdchen since you are not updating a local branch, and you are checking out a specific commit in the `_reset_git_head` method, you don't need to pull. You just need to get the remote metadata with `git fetch`.
This snippet illustrates what I'm trying to say:
```
$ mkdir temp-cachito && cd temp-cachito
$ git init
Initialized empty Git repository in /tmp/temp-cachito/.git/
$ git remote add origin https://github.com/release-engineering/cachito.git
$ git fetch origin
remote: Enumerating objects: 61, done.
remote: Counting objects: 100% (61/61), done.
remote: Compressing objects: 100% (51/51), done.
remote: Total 1338 (delta 32), reused 16 (delta 10), pack-reused 1277
Receiving objects: 100% (1338/1338), 302.66 KiB | 1.78 MiB/s, done.
Resolving deltas: 100% (896/896), done.
From https://github.com/release-engineering/cachito
 * [new branch]      document-invalid-query-param -> origin/document-invalid-query-param
 * [new branch]      gh-pages                     -> origin/gh-pages
 * [new branch]      master                       -> origin/master
 * [new branch]      remove-cachito-cli           -> origin/remove-cachito-cli
 * [new branch]      upgrade-athens               -> origin/upgrade-athens
$ git checkout d34406a99d5a5573823b6c9eb58a62318950ef5b
Note: checking out 'd34406a99d5a5573823b6c9eb58a62318950ef5b'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at d34406a Do not run flake8 on .env dir
```

Reply:
@mprahl Changed to use `fetch`. The `_reset_git_head` uses `git-reset`, and I tested the `git-reset` also works without a precedent merge.